### PR TITLE
[1386] Allow users to upload their own custom images

### LINKED ISSUE #1386: Allow users to upload their own custom images
Currently the only way to add custom images to the backend so that they can be used is via APIs (see `org.eclipse.sirius.web.sample.configuration.CustomImagesLoader` for example).

It should be possible for end-users to upload their own custom images via the web interface. The resulting images should then automatically be usable:
* in diagrams as the image for an `ImageNodeStyle`;
* in forms via the `Image` widget (see #1381).

### PR DESCRIPTION
**Pull request by pcdavid**
Signed-off-by: Pierre-Charles David <pierre-charles.david@obeo.fr>

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [x] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [x] Does the title of the commit contributing the ADR start with `[doc]`?
- [x] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/074_custom_image_upload.adoc [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+type DeleteImageSuccessPayload {
+  id: ID!
+}
+
+union DeleteImagePayload = ErrorPayload | DeleteImageSuccessPayload
+----
+
+The
```
Comment (AxelRICHARD) [inline doc/adrs/074_custom_image_upload.adoc]: Please remove this line

**Review by AxelRICHARD — CHANGES_REQUESTED** [review]
Ok on my side for this ADR

File: doc/adrs/074_custom_image_upload.adoc (hunk 1 above)
Comment (pcdavid) [reply doc/adrs/074_custom_image_upload.adoc]: It's not finished, I pushed too early; it's missing the description of the new GraphQL query needed to get the images's list to present in the UI.

File: packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/mutation/MutationDeleteImageDataFetcher.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
```
Comment (AxelRICHARD) [inline packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/mutation/MutationDeleteImageDataFetcher.java:1]: Please add a CHANGELOG entry for this commit/ticket

File: packages/view/backend/sirius-components-view-edit/src/main/java/org/eclipse/sirius/components/view/provider/ViewItemProviderAdapterFactory.java [hunk 3]
```diff
@@ -1578,6 +1578,29 @@ public Adapter createButtonDescriptionAdapter() {
         return this.buttonDescriptionItemProvider;
     }
 
+    /**
```
Comment (AxelRICHARD) [inline packages/view/backend/sirius-components-view-edit/src/main/java/org/eclipse/sirius/components/view/provider/ViewItemProviderAdapterFactory.java]: Please add new ImageDescription in org.eclipse.sirius.components.view.provider.FormDescriptionItemProvider.collectNewChildDescriptors(Collection<Object>, Object)

**Review by AxelRICHARD — CHANGES_REQUESTED** [review]

File: packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/mutation/MutationDeleteImageDataFetcher.java (hunk 2 above)
Comment (pcdavid) [reply packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/mutation/MutationDeleteImageDataFetcher.java:1]: Done.

File: packages/view/backend/sirius-components-view-edit/src/main/java/org/eclipse/sirius/components/view/provider/ViewItemProviderAdapterFactory.java (hunk 3 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-edit/src/main/java/org/eclipse/sirius/components/view/provider/ViewItemProviderAdapterFactory.java]: Done in https://github.com/eclipse-sirius/sirius-components/pull/1381/files#diff-acd82297577acd07c190a16d7afbbbe9deb19e2d525d625d8612346064dafaa7

**Review by AxelRICHARD — APPROVED** [review]

**Pull request merged**